includes/constants/class-base-constant.php

```diff
@@ -0,0 +1,106 @@
+<?php
+/**
+ * Class Base_Constant
+ *
+ * @package WooCommerce\Payments
+ */
+
+namespace WCPay\Constants;
+
+use ReflectionClass;
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit; // Exit if accessed directly.
+}
+
+
+/**
+ * Base constant class to hold common logic for all constants.
+ */
+abstract class Base_Constant {
+
+	/**
+	 * Enum value
+	 *
+	 * @var mixed
+	 */
+	protected $value;
+
+	/**
+	 * Class constructor.
+	 *
+	 * @param mixed $value Constant from class.
+	 * @throws \InvalidArgumentException
+	 */
+	public function __construct( $value ) {
+		if ( $value instanceof static ) {
+			$value = $value->get_value();
+		} else {
+			if ( ! defined( static::class . "::$value" ) ) {
+				throw new \InvalidArgumentException( "Constant with name '$value' does not exist." );
+			}
+		}
+
+		$this->value = $value;
+	}
+
+	/**
+	 * Get enum class value.
+	 *
+	 * @return mixed
+	 */
+	public function get_value() {
+		return $this->value;
+	}
+
+	/**
+	 * Compare to enums.
+	 *
+	 * @param mixed $variable Constant object to compare.
+	 *
+	 * @return bool
+	 */
+	final public function equals( $variable = null ): bool {
+		return $this->get_value() === $variable->get_value() && static::class === \get_class( $variable );
```

Comment:
```suggestion
		return is_object( $variable ) && static::class === \get_class( $variable ) && $this->get_value() === $variable->get_value();
```

Tight up the conditions here: move the check class name as the second condition and a check with `is_object`? Otherwise, comparing with a random value (not created by this Enum) can yell errors. Then add the relevant tests too. 

You can apply this diff for the test file https://gist.github.com/htdat/c874e552412fc2c8fd85a33b0c91295a and run PHP tests to see these errors ` npm run test:php -- --filter=Base_Constant_Test`.

Reply:
`is_object` validation is not necessary because get_class will return the name of the class, even for primitive types like int and string. The only thing that we could check is that the `$variable` is an instance of a base constant, so we don't have an issue with get_value() not existing on the passed variable.

Reply:
> because get_class will return the name of the class, even for primitive types like int and string

This seems not... true. https://3v4l.org/tFM3p

> The only thing that we could check is that the $variable is an instance of a base constant, so we don't have an issue with get_value() not existing on the passed variable.

I think this works too.

Reply:
Fixed with the latest commit